runner.py

```diff
@@ -0,0 +1,707 @@
+    RESULT = r"TEST\((\w+), (\w+)\) (PASS|IGNORE)"
+    # Fail need to have its own regex due to greedy matching
+    RESULT_FAIL = r"TEST\((\w+), (\w+)\) (FAIL) at (.*):(\d+)\r"
+    FINAL = r"(\d+) Tests (\d+) Failures (\d+) Ignored"
+
+    @staticmethod
+    def harness(proc):
+        test_results = []
+        last_assertion = {}
+
+        while True:
+            idx = proc.expect([
+                UnitTestHarness.ASSERT,
+                UnitTestHarness.RESULT,
+                UnitTestHarness.RESULT_FAIL,
+                UnitTestHarness.FINAL
+            ])
+            groups = proc.match.groups()
+
+            if idx == 0:
+                assertion = dict(zip(('path', 'line', 'status', 'msg'), groups))
+                # We only care for fail messages
+                if assertion['status'] == 'FAIL':
+                    last_assertion = assertion
+            elif idx in (1, 2):
+                test = dict(zip(('group', 'name', 'status'), groups[:3]))
+
+                # If fail then match info with last assertion
+                if test['status'] == 'FAIL':
+                    test.update(dict(zip(('path', 'line'), groups[3:])))
+                    if (test['path'] in last_assertion.values()
+                       and test['line'] in last_assertion.values()):
+                        test['msg'] = last_assertion['msg']
+
+                test_results.append(UnitTest(**test))
+            elif idx == 3:
+                test_stats = list(map(int, groups))
+                total, fail, ignore = test_stats
+
+                # Make sure if parsed infos are as expected
+                fail_no = len([t for t in test_results if t.status == UnitTest.FAIL])
+                ignore_no = len([t for t in test_results if t.status == UnitTest.IGNORE])
+                if (fail_no != fail or ignore_no != ignore or len(test_results) != total):
+                    raise ValueError("Unit test harness: results do not match")
+
+                return test_results
+
+
+class TestCase:
+    PASS = "PASS"
+    FAIL = "FAIL"
+    FAIL_TIMEOUT = "FAIL TIMEOUT"
+    IGNORE = "IGNORE"
+
+    def __init__(self, name, target, syspage=None, status=None):
+        self.name = name
+        self.target = target
+        self.syspage = syspage
+        if not status:
+            status = TestCase.FAIL
+        self.status = status
+        self.harness = None
+
+    def __str__(self):
+        status = self.status
+        if status == TestCase.PASS:
+            color = Color.OK
+        elif status in (TestCase.FAIL, TestCase.FAIL_TIMEOUT):
+            color = Color.FAIL
+        elif status == TestCase.IGNORE:
+            color = Color.IGNORE
+
+        status = Color.colorify(self.status, color)
+        return f"{self.target}: {self.name}: {status}"
+
+    def ignored(self):
+        return self.status == TestCase.IGNORE
+
+    def failed(self):
+        return self.status == TestCase.FAIL
```

Comment:
or `TestCase.FAIL_TIMEOUT`

Right now timeouted tests are considered not an error.